test/common/bidi.py

```diff
@@ -0,0 +1,438 @@
+    if _type == "object":
+        obj = {}
+        for k, v in raw["value"]:
+            obj[k] = unpack_value(v["value"])
+        return obj
+    raise ValueError(f"Unknown type in {raw!r}")
+
+
+class WebdriverBidi:
+    http_session: aiohttp.ClientSession
+
+    def __init__(self, headless: bool = False) -> None:
+        self.headless = headless
+        self.last_id = 0
+        self.pending_commands: dict[int, asyncio.Future] = {}
+        self.logs: list[LogMessage] = []
+        self.bidi_session: BidiSession | None = None
+        self.future_wait_page_load = None
+        self.top_context: str | None = None  # top-level browsingContext
+        self.context: str | None  # currently selected context (top or iframe)
+
+        self.homedir = tempfile.TemporaryDirectory(prefix="browser-home-")
+        self.download_dir = Path(self.homedir.name) / 'Downloads'
+        self.download_dir.mkdir()
+
+    async def start_bidi_session(self) -> None:
+        raise NotImplementedError('must be implemented by concrete subclass')
+
+    async def close_bidi_session(self) -> None:
+        raise NotImplementedError('must be implemented by concrete subclass')
+
+    async def close(self) -> None:
+        assert self.bidi_session is not None
+        log_proto.debug("cleaning up webdriver")
+
+        self.task_reader.cancel()
+        del self.task_reader
+        await self.ws.close()
+        await self.close_bidi_session()
+        self.bidi_session.process.terminate()
+        await self.bidi_session.process.wait()
+        self.bidi_session = None
+        await self.http_session.close()
+
+    def ws_done_callback(self, future):
+        for fut in self.pending_commands.values():
+            if not fut.done():
+                fut.set_exception(WebdriverError("websocket closed"))
+        if not future.cancelled():
+            log_proto.error("ws_reader crashed: %r", future.result())
+
+    async def start_session(self) -> None:
+        self.http_session = aiohttp.ClientSession(raise_for_status=True)
+        await self.start_bidi_session()
+        assert self.bidi_session
+        self.ws = await self.http_session.ws_connect(self.bidi_session.ws_url)
+        self.task_reader = asyncio.create_task(self.ws_reader(self.ws), name="bidi_reader")
+        self.task_reader.add_done_callback(self.ws_done_callback)
+
+        await self.bidi("session.subscribe", events=[
+            "log.entryAdded", "browsingContext.domContentLoaded",
+        ])
+
+        # wait for browser to initialize default context
+        for _ in range(10):
+            realms = (await self.bidi("script.getRealms"))["realms"]
+            if len(realms) > 0:
+                self.top_context = realms[0]["context"]
+                self.context = self.top_context
+                break
+            await asyncio.sleep(0.5)
+        else:
+            raise WebdriverError("timed out waiting for default realm")
+
+    async def __aenter__(self):
```

Comment:
Oh.  This is already a context manager.

Reply:
In that case you should subclass from `AbstractAsyncContextManager` to declare that you implement this interface.

Reply:
Yes, but it's a leftover, I don't use this anywhere as a CM. I left it in in case you wanted to use it for the unit tests.

I made it a subclass, but mypy complains:

```
class WebdriverBidi(contextlib.AbstractAsyncContextManager):     ■ Missing type parameters for generic type "AbstractAsyncContextManager"  [type-arg]
```

I have no idea what it wants from me, and https://docs.python.org/3/reference/datamodel.html#async-context-managers doesn't explain any types either. That's the "total waste of time" aspect of strict type checking :cry: I ignored that for now, perhaps you know what to put in here.

Reply:
Conceptually you want `[Self]`, but I'm not sure if that's valid when subclassing and I'm also not sure if that's available in older versions of Python (in fact, I think it's not)...

Reply:
Yeah, neitner `contextlib.AbstractAsyncContextManager['Self']` nor `contextlib.AbstractAsyncContextManager[Self]` work unfortunately..